pydarn/io/borealis/restructure_borealis.py

```diff
@@ -0,0 +1,774 @@
+            # which is why a num_beams array is necessary.
+            beam_nums_array[rec_idx][0:len(beam_nums_buffer)] = beam_nums_buffer 
+            beam_azms_array[rec_idx][0:len(beam_azms_buffer)] = beam_azms_buffer
+            num_beams_array[rec_idx] = len(beam_nums_buffer)
+
+            # insert data into buffer
+            record_buffer = data_dict[k]["data"]
+            data_pos = rec_idx * data_buffer_offset
+            data_end = data_pos + len(record_buffer)
+            data_buffer[data_pos:data_end] = record_buffer
+
+            # insert sequence timestamps into buffer
+            rec_sqn_ts = data_dict[k]["sqn_timestamps"]
+            sqn_pos = rec_idx * sqn_buffer_offset
+            sqn_end = sqn_pos + data_dict[k]["num_sequences"]
+            sqn_ts_buffer[sqn_pos:sqn_end] = rec_sqn_ts
+
+            rec_noise = data_dict[k]["noise_at_freq"]
+            noise_pos = rec_idx * noise_buffer_offset
+            noise_end = noise_pos + data_dict[k]["num_sequences"]
+            noise_buffer[noise_pos:noise_end] = rec_noise
+
+        # write leftover metadata and data
+        new_data_dict["num_sequences"] = sqn_num_array
+        new_data_dict["int_time"] = int_time_array
+        new_data_dict["scan_start_marker"] = scan_start_array
+        new_data_dict["num_slices"] = num_slices_array
+        new_data_dict["beam_nums"] = beam_nums_array 
+        new_data_dict["beam_azms"] = beam_azms_array
+        new_data_dict["num_beams"] = num_beams_array
+
+        new_data_dict["data"] = data_buffer.reshape(data_shape)
+        new_data_dict["sqn_timestamps"] = sqn_ts_buffer.reshape(sqn_shape)
+        new_data_dict["noise_at_freq"] = noise_buffer.reshape(noise_shape)
+
+        new_data_dict["data_descriptors"] = data_descriptors
+        # new_data_dict["data_dimensions"] = data_dims_array
+
+        return new_data_dict
+
+    @classmethod
+    def bfiq_site_to_array(cls, filename: str, data_dict: OrderedDict) -> dict:
+        """
+        Restructuring method for bfiq data.
+
+        Parameters
+        ----------
+        filename
+            String of where file originated from for better error messages.
+        data_dict
+            a dict of timestamped records loaded from an hdf5 Borealis bfiq
+            data file
+        
+        Returns
+        -------
+        new_data_dict
+            A dictionary containing the data from data_dict
+            reformatted to be stored entirely as arrays, or as
+            one entry if the field does not change between records
+        """
+        
+        try:
+            new_data_dict = cls._iq_site_to_array(data_dict, 'bfiq')
+
+            BorealisUtilities.check_arrays(filename, new_data_dict, 
+                borealis_formats.BorealisBfiq.array_single_element_types(),
+                borealis_formats.BorealisBfiq.array_array_dtypes(),
+                borealis_formats.BorealisBfiq.unshared_fields)
+
+        except Exception as e:
```

Comment:
I want to catch all exceptions here and redirect to restructure error with traceback so that the user knows the error is coming from the restructuring code and not just read/write code. Also, so that the error can be caught as a restructure error specifically

Reply:
I am confused, normally when an exception is thrown a traceback will happen and you will see the hierarchy of calls the induced the exception so you will know if it came from restructuring vs. read/write. Usually why PEP8  discourages wrapping all your code in a try/except block 
> Additionally, for all try/except clauses, limit the try clause to the absolute minimum amount of code necessary. Again, this avoids masking bugs.

If the latter of your reasoning is the stronger reasoning you actually can avoid a lot of this syntax using `raise from` which is in python 3 - https://franklingu.github.io/programming/2016/06/30/properly-reraise-exception-in-python/

Generally, I am not a fan of wrapping all exceptions into another exception just in case a user wants to catch a specific exception for passing but not the others. However, we can play that by ear as well.

Reply:
I found an example on pep8.org that I think is very similar to this case. I think the traceback or 'raise from' syntax will avoid masking bugs. I will change to raise from because it looks much cleaner.

>When binding caught exceptions to a name, prefer the explicit name binding syntax added in Python 2.6:
try: 
     process_data()  
except Exception as exc:  
    raise DataProcessingFailedError(str(exc))

I also found this:
>Design exception hierarchies based on the distinctions that code catching the exceptions is likely to need, rather than the locations where the exceptions are raised. Aim to answer the question “What went wrong?” programmatically, rather than only stating that “A problem occurred”

I think I'm doing this but not sure. I think relabelling as a restructure error is appropriate for catching by other code which should (I hope) identify something wrong with the data.

Reply:
I have no issue with renaming giving it provides the traceback to allow the user to know "what went wrong" because "NameError" "ValueError" and "OSError" are rather important to know in the case of what went wrong with the data and to allow us for debugging if there is a bug in the code. 

If it makes for try/except easier in your higher-level code in pyDARN that is fine, just make sure you explain so others understand why you need to rename it. Usually, re-raising is not common unless you need to do something before raising an exception.

Reply:
resolve conversation, when you implement `raise from`.